app/uk/gov/hmrc/personaldetailsvalidation/audit/AuditDataEventFactory.scala

```diff
@@ -43,12 +43,24 @@ private[personaldetailsvalidation] class AuditDataEventFactory(auditConfig: Audi
                       (implicit hc: HeaderCarrier, request: Request[_]): DataEvent = createEvent(personalDetails, "technicalError")
 
   private def createEvent(personalDetails: PersonalDetails, matchingStatus: String)
-                         (implicit hc: HeaderCarrier, request: Request[_]): DataEvent = DataEvent(
-    auditSource = auditConfig.appName,
-    auditType = auditType,
-    tags = auditTagProvider(hc, auditType, request),
-    detail = auditDetailsProvider(hc) + ("nino" -> personalDetails.nino.value) + ("matchingStatus" -> matchingStatus)
-  )
+                         (implicit hc: HeaderCarrier, request: Request[_]): DataEvent = {
+    val nino = personalDetails.nino match {
+      case Some(ninoValue) => ninoValue.value
+      case None => "NOT SUPPLIED"
+    }
+
+    val postCode = personalDetails.postCode match {
+      case Some(postCodeValue) => postCodeValue
+      case None => "NOT SUPPLIED"
+    }
+
+    DataEvent(
+      auditSource = auditConfig.appName,
+      auditType = auditType,
+      tags = auditTagProvider(hc, auditType, request),
+      detail = auditDetailsProvider(hc) + ("nino" -> nino) + ("postCode" -> postCode) + ("matchingStatus" -> matchingStatus)
```

Comment:
We don't need to audit both the nino and postcode for the same validation request. We know it is going to be either of the cases.

Reply:
I guess once we go for the ADT solution Anshul mentioned below, we could have two different strategies implicitly injected into the 'createEvent' method to cope with both of the types.